### PR TITLE
[diffusion] model: support FastH3 (4-step VSA-distilled MiniMax-H3) with a VSA-H3 attention backend

### PR DESCRIPTION
**Pull request by kevin-mii**
## Motivation

Add [FastVideo/FastVideo-FastH3-4-step-Preview-v1-VSA-DataFree](https://huggingface.co/FastVideo/FastVideo-FastH3-4-step-Preview-v1-VSA-DataFree), the 4-step DMD2 distillation of MiniMax-H3 trained with Video Sparse Attention (0.9 sparsity, 64-token tiles), together with the `video_sparse_attn_h3` backend it was trained for.

4x B300, 1344x768 at 24 fps with audio, bf16 eager, Ulysses 4, `VSA_sparsity` 0.9, seed 1000. Clean E2E is the client wall clock of a `/v1/videos` request against `sglang serve` including decode, muxing and file output (median of three after one warm request); the stage columns are the server timings of the same request. H3 aligns the requested durations to 124, 243 and 362 frames.

| Requested / aligned | Encoder | DiT total / per forward | Video + audio VAE | Transport + MP4 | Clean E2E | Client RTF | x real time | Peak HBM |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 5 s / 124 | 0.07 s | 2.18 s / 0.545 s | 0.87 s | 0.9 s | **4.1 s** | 0.79 | 1.26 | 94 GB/GPU |
| 10 s / 243 | 0.07 s | 4.83 s / 1.21 s | 1.71 s | 1.3 s | **8.0 s** | 0.79 | 1.27 | 103 GB/GPU |
| 15 s / 362 | 0.07 s | 8.80 s / 2.20 s | 2.56 s | 1.8 s | **13.3 s** | 0.88 | 1.13 | 111 GB/GPU |

All three satisfy client RTF <= 1.0. For reference, FastVideo's published 4x B200 VSA numbers for the same shapes are 6.1 / 12.0 / 15.5 s, and vllm-project/vllm-omni#6909 reports 8.68 s clean E2E at 10 s and 14.18 s at 15 s on 8x B300 (1.383 s and 2.379 s per forward). Dense FA4 on the same weights and GPUs: 3.77 / 9.84 / 18.45 s (`sglang generate`, stage sum).

## Modifications

- **Model.** `FastH3PipelineConfig` / `FastH3SamplingParams` / `FastH3Pipeline` are thin MiniMax-H3 variants that reuse every H3 stage, loader and admission path; no core files change. The flat native-Diffusers repo is materialized into the base-H3 layout by a revision-pinned model overlay (`kevin-mi/FastH3-4step-Preview-overlay@f769cb80`, same mechanism as the LTX-2.3 / JoyAI / SANA-WM entries). The distilled schedule is the standard shift-12/3 sigma grid at `num_inference_steps=5` (four DiT forwards); no DMD plumbing. `has_gate_compress` adds the trained per-block `to_gate_compress` projection, kept bf16 and unquantized; base H3 loads zeros and runs pure sparse.
- **VSA-H3 backend** (`--attention-backend video_sparse_attn_h3`). Segment-pure prefix tiles plus (4, 4, 4) video tiles over H3's packed `[text | keyframes | audio | video]` sequence; the FastVideo tile-64 Triton kernel with TMA loads and a pinned launch config; one fused pack-and-pool kernel, top-k emitted directly as kernel index lists, one fused untile with the compression branch folded in. Knobs: `VSA_sparsity`, `vsa_mode` (`exempt` / `compete`), `vsa_dense_first_n_steps`, `vsa_dense_layers`. Only the DiT runs sparse; the token refiner, text encoder and VAEs keep their dense defaults. Ulysses supported; ring, `torch.compile` and breakable CUDA graph rejected. Validated on SM90, SM100 and SM103.
- **Rejections.** `--model-variant`, `quality="high"`, `fl2va` / `ref2va`, any `num_inference_steps` other than 5, `vsa_tile_size != 64`, a checkpoint missing any trained `to_gate_compress.weight` (required parameter, never synthesized), and the sibling FastH3 LoRA bundles (their `.diff` / `.diff_b` / `set_weight` tensors match no LoRA mapping rule and would otherwise be dropped silently).
- **Tests and docs.** `test_minimax_h3_fasth3.py` (5), `test_vsa_h3_attention.py` (6, including a sparsity-0.5 masked-dense reference with gates for both modes), GPU case `fasth3_t2va_vsa_4gpu_h100`, cookbook section, attention-backends and compatibility-matrix entries, `fasth3-t2va-vsa` benchmark preset.

Validated deployments on 4x B300 at 5 s: tp2 x ulysses2 (3.42 s, 61 GB), `--use-fsdp-inference` (3.42 s, 50 GB), `--quantization fp8` (2.93 s, 63 GB, gates stay bf16), `--quantization kitchen_int8` (functional), `num_outputs_per_prompt` fan-out, `torch_sdpa` and `sage_attn` dense fallbacks. Rejected with a clear message: `subblock_sparse_attn` on SM103 (sm_100a-only kernel), the AdaLN cache (requires `--model-variant`), and the LoRA bundle. `SGLANG_CACHE_DIT_ENABLED=1` behaves as on FastWan: mounts, never hits at four forwards, output unchanged.

## Accuracy Tests

The port follows FastVideo's `video_sparse_attn_h3` backend and `block_sparse_attn_triton.py`: same tile geometry, pooling, top-k budgets (`exempt` / `compete`), compression branch and online-softmax kernel. On 4x H200 the VSA-H3 output was checked frame by frame against the FastVideo upstream reference for the same prompt and seed and judged at parity. The fused data path is bit-identical to the original port without the gate and within one bf16 ulp with it; the unit test compares the kernel against an fp32 masked-dense reference at sparsity 0.5 with gates in both modes.

4x B300, 1344x768, 5 s, seed 1000, t = 0.5 / 2.5 / 4.5 s:

VSA-H3 frames

Note for reviewers: the 4-step schedule amplifies rounding-level differences (Ulysses vs TP2 x Ulysses2, fp8, one-ulp kernel changes) into a different but equally coherent sample from the same seed, so SSIM against a single-seed reference is not a meaningful gate for this model.

## Speed Tests and Profiling

`compare_perf.py`, dense FA4 baseline vs VSA-H3 on the same weights (`sglang generate`, 4x B300, 2-step warmup):

| Video | Dense FA4 E2E | VSA-H3 E2E | Diff |
| --- | ---: | ---: | ---: |
| 5 s | 3765 ms | 3183 ms | -15.5% |
| 10 s | 9844 ms | 6683 ms | -32.1% |
| 15 s | 18454 ms | 11471 ms | -37.8% |

Profile-driven: the first trace showed the sparse kernel at 10 ms per DiT layer next to 12 ms of tile gather/scatter, stack, transposes and mask-to-index; the fused path brings that bookkeeping to about 1 ms per layer (35 to 24 ms per layer at 10 s). Remaining per-layer budget at 10 s: attention 9.0 ms, GEMMs 8.4 ms, all-to-all 2.5 ms. The kernel launch config is pinned from a 12-config sweep on B300 (`num_warps=4`, `num_stages=5`), so a fresh process pays 0.3 s on its first sparse call instead of a 2.3 s autotune per sequence length.

## Known limitations

- t2va only; fl2va / ref2va were not distilled.
- Upstream labels the checkpoint a preview; quality gaps versus base H3 on hard motion and fine detail are properties of the distillation.
- The sibling `FastVideo-FastH3-4-step-Preview-v1-LoRA` adapters are out of scope by design: fusing them onto base H3 reproduces the merged checkpoint this PR serves.
- **License note for maintainers:** the checkpoint inherits the MiniMax-H3 Community License. FastH3 is listed where FastWan is (README support line, compatibility matrix) plus the MiniMax-H3 cookbook section; drop it from those lists if the license needs review first.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #33629244466](https://github.com/sgl-project/sglang/actions/runs/33629244466)
Latest PR Test (Extra): :x: [Run #33629244087](https://github.com/sgl-project/sglang/actions/runs/33629244087)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #33629244460](https://github.com/sgl-project/sglang/actions/runs/33629244460)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py [hunk 1]
```diff
@@ -757,6 +794,19 @@ def __init__(
...
+            self.to_gate_compress = ColumnParallelLinear(
+                arch.hidden_size,
+                self.inner_dim,
+                bias=False,
+                gather_output=False,
+                params_dtype=_BF16_DTYPE,
+                quant_config=None,
+                prefix=f"{prefix}.to_gate_compress",
+            )
+            self.to_gate_compress.weight.missing_param_init = "zeros"
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py]: [P1] Please fail closed when a trained FastH3 gate is missing. `MiniMaxH3DiTModel.__init__()` later marks every parameter without an explicit policy as `missing_param_init = "error"`, but this explicit `"zeros"` survives that pass, and `fsdp_load.py` then synthesizes a zero tensor. Base H3 already represents the no-gate behavior with `has_gate_compress=False`; whenever this module exists, the FastH3 checkpoint is expected to provide each trained `to_gate_compress.weight` (the released checkpoint has one for every main block). An incomplete/broken overlay would therefore start successfully as a different pure-sparse model and silently change output quality. Please use the required/error policy here and add a missing-gate loading test.

File: python/sglang/multimodal_gen/configs/sample/minimax_h3.py [hunk 2]
```diff
@@ -302,4 +302,20 @@ def refresh_request_extra_after_output_expansion(self, req: Any) -> None:
...
 
 
-__all__ = ["MiniMaxH3SamplingParams"]
+@dataclass
+class FastH3SamplingParams(MiniMaxH3SamplingParams):
+    """FastH3: five sigma grid points, i.e. the four distilled DiT forwards."""
+
+    num_inference_steps: int = 5
+
+    def _validate(self) -> None:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/configs/sample/minimax_h3.py:311]: [P1] Please enforce `num_inference_steps == 5` in the FastH3 admission contract. This class documents five sigma points/four DiT evaluations, the released inference metadata pins the four DMD steps `[999, 749, 500, 250]`, and the cookbook says not to send the base schedule. Today `FastH3SamplingParams(..., num_inference_steps=50)` passes this validator; timestep preparation then creates a 50-point shifted grid and runs 49 off-distribution DiT forwards instead of rejecting it. That is a silent correctness and latency failure for an otherwise accepted request. Please reject non-5 values here and add a negative test.

**Review by BBuf — CHANGES_REQUESTED** [review]
Change summary: this PR adds the FastH3 model/overlay and its fixed distilled sampling path, the learned `to_gate_compress` projection, a `video_sparse_attn_h3` backend with an in-tree Triton/TMA block-sparse kernel, per-step denoising metadata, Ulysses support, docs, benchmarks, and unit/server coverage. The VSA math and tensor layout are broadly aligned with the FastVideo reference, but two model-contract paths currently fail open, so I am requesting changes.

```mermaid
flowchart LR
    A["/v1/videos T2VA request"] --> B["FastH3 registry + pinned overlay"]
    B --> C["FastH3 sampling/admission"]
    C --> D["Per-step VSA-H3 metadata"]
    D --> E["MiniMax-H3 Q/K/V + trained gate"]
    E --> F["Ulysses all-to-all"]
    F --> G["64-token tile pack/pool + top-k"]
    G --> H["Triton sparse attention + compression"]
    H --> I["Untile + DiT/VAEs"]
    L["Legend: dashed border = changed in this PR"]
    classDef changed stroke-dasharray: 5 5
    class B,C,D,E,F,G,H,L changed
```

The registry recognizes the released or materialized FastH3 repository, applies the pinned base-H3 overlay, and enables the per-block compression gate. Admission then selects the five-point shifted sigma schedule. On every DiT step, the denoising stage installs segment/tile metadata; the attention layer sequence-shards Q/K/V and the gate under Ulysses, pools 64-token tiles, chooses the trained top-k policy, executes sparse attention, adds the gated pooled-value correction, and scatters the result back into packed-token order before the normal H3 output and VAE paths continue.

Historical coverage: I swept all 26 touched paths plus the relevant kernel/model-loading/sparse-attention risks through the SGLang maintainer-review corpus, then widened the pass to non-inline PR conversations. The available MiniMax optimization history is MiniMax-M2-only and has no direct H3/diffusion precedent, so I used the FastVideo implementation, the released FastH3 inference metadata, and the published checkpoint contents as the semantic references. The recurring relevant review expectation is that new model/kernel paths encode their checkpoint and schedule assumptions explicitly and exercise the claimed GPU/topology path.

Findings:

- **[P1] Require every trained FastH3 gate weight** — `python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py:809`. The explicit zero-init policy bypasses the model-wide required-parameter pass, so an incomplete overlay can load successfully as a different model.
- **[P1] Reject non-distilled inference-step counts** — `python/sglang/multimodal_gen/configs/sample/minimax_h3.py:311`. The accepted value needs to be exactly five sigma points; otherwise the code generates and runs an off-distribution schedule instead of rejecting the request.

Validation performed: `git diff --check`, `compileall`, and the CI lint selection (`F401,F821,UP037`) pass on this head. I could not execute the Torch/pytest or GPU suites in the local environment. GitHub lint passed, but the base/extra GPU workflows stopped at `pr-gate` because the PR has no `run-ci` label, so the newly registered 4xH100 case and the VSA tests have not run in CI. Please run the relevant CI after the fixes. Also note that the H100 server case currently disables consistency, component-accuracy, and input-reference checks, so it is only a generation smoke test; the stated 4xH200 frame parity and B300 measurements remain manual evidence, especially for the SM103 path.

File: python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py (hunk 1 above)
Comment (kevin-mii) [reply python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py]: Fixed in bfbbd2d0b4: the explicit `missing_param_init = "zeros"` is removed, so `to_gate_compress.weight` falls under the model-wide required policy set by `_mark_missing_params_required()` and a checkpoint missing any gate now fails the load with the required-parameter error. Base H3 keeps the no-gate behaviour through `has_gate_compress=False` (no module is built). `test_fasth3_gates_stay_bf16_under_runtime_quantization` asserts the gate carries the `error` policy on the FastH3 meta model. Re-verified the released checkpoint still loads (all 50 gates present) with byte-identical output.

File: python/sglang/multimodal_gen/configs/sample/minimax_h3.py (hunk 2 above)
Comment (kevin-mii) [reply python/sglang/multimodal_gen/configs/sample/minimax_h3.py:311]: Fixed in bfbbd2d0b4: `FastH3SamplingParams._validate` now rejects any `num_inference_steps` other than 5 with a message pointing to base MiniMax-H3 for other schedules; `test_fasth3_sampling_defaults_and_task_rejection` covers `num_inference_steps=50`. Warmup requests are unaffected since `Req.set_as_warmup` overrides the step count after admission. Cookbook wording updated to match. The branch is also rebased onto current main (docs conflict with the new SpargeAttention section resolved).

**Review by BBuf — APPROVED** [review]

**Review by zijiexia — APPROVED** [review]
